### ISSUE TITLE
External Storage SFTP public-key authentication setup is misleading and the “Generate keys” option is missing

### ISSUE THREAD

**Issue by HannesJo0139** [nextcloud/server#35829]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Trying to mount External Storage using SFTP and Public Key authentication in the user settings. The doc https://docs.nextcloud.com/server/25/admin_manual/configuration_files/external_storage/auth_mechanisms.html states that a key pair is generated by nextcloud. The Screenshot in the doc shows a _generate keys_ button. 

- However, there actually is no such button. Instead, it wants me to enter a public key. 
- Also the text description seems misplaced. It says `... You may also allow users to mount their own external storage services.` what is obviously an information supposed to be in the admin settings, but it is in the user settings of a non-admin account.

![image](https://user-images.githubusercontent.com/42114183/208470225-87391b9c-2109-43e7-943c-e6acf536b2a6.png)

### Steps to reproduce

1. Setup NC
2. Activate External storage
3. Go to user settings -> External Storage

### Expected behavior

Behavior should match what is described in the docs or the docs should be updated. Also Non-admin users should not see admin text description.

### Installation method

None

### Operating system

None

### PHP engine version

None

### Web server

None

### Database engine version

None

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### List of activated Apps

```shell
External Storage
```

**Comment by szaimen** [comment]
which NC version?

**Comment by HannesJo0139** [comment]
25.0.1

**Comment by HannesJo0139** [comment]
Seems like people using dirty workarounds for it for several months: https://www.reddit.com/r/NextCloud/comments/xcqam0/external_storage_sftp_with_rsa_generate_key/

**Comment by kesselb** [comment]
I believe some code for the "generate keys" feature is still there but the button is not. 

https://github.com/nextcloud/server/blob/f5b833845c0c183d1ef9b8d41282bbceda45e0a7/apps/files_external/js/public_key.js#L3

https://github.com/nextcloud/server/blob/215aef3cbdc1963be1bb6bca5218ee0a4b7f1665/apps/files_external/lib/Controller/AjaxController.php#L88

**Comment by HannesJo0139** [comment]
I mean I could also generate a key pair myself and use the private key option? But currently the public key option is useless if I got it right. I am kind of missing some info about the options or maybe I got sth wrong, but client would always need a private key in order to use pubkey auth, right?

**Comment by xundeenergie** [comment]
On my destination-host i get this errormessage

`Unable to negotiate with 0.0.0.0 port 42520: no matching host key type found. Their offer: rsa-sha2-256,rsa-sha2-512,ssh-rsa,ssh-dss [preauth]`

So it seems, nextcloud does not use local installed sshd. Because i can connect with my privatekey via sftp to the target-host from commandline.

**Comment by HannesJo0139** [comment]
Seems hopeless. Rolling-out new features all day long but not maintaining 'em properly. At Nextcloud, quantity goes before quality. Very disappointing..

**Comment by kesselb** [comment]
> So it seems, nextcloud does not use local installed sshd. Because i can connect with my privatekey via sftp to the target-host from commandline.

Nextcloud does not use the ssh agent. You need to add the private key for the remote sftp server to Nextcloud as explained in the referenced Reddit thread. 

> Seems hopeless.

Not hopeless, but also not our top priority. 
The approach to generate the key pairs yourself is fine. The generate keys button is a shortcut.

**Comment by kesselb** [comment]
I'm a developer. I don't do marketing, I don't do roadmaps, and therefore cannot discuss marketing or roadmap related concerns. 

If you have some feedback for our marketing people, try to reach out via Twitter/Mastodon/email or stop by the office.

If you have a support contract with Nextcloud GmbH, please report this problem via customer portal or email.

**Comment by xundeenergie** [comment]
> > So it seems, nextcloud does not use local installed sshd. Because i can connect with my privatekey via sftp to the target-host from commandline.
> 
> Nextcloud does not use the ssh agent. You need to add the private key for the remote sftp server to Nextcloud as explained in the referenced Reddit thread.

I created such a keypair and inserted them...

It's the hostkey, not the userkey.

So i created an rsa-hostkey on both machines too... but i still get this error.

**Comment by HannesJo0139** [comment]
> I'm a developer. I don't do marketing, I don't do roadmaps, and therefore cannot discuss marketing or roadmap related concerns.
> 
> If you have some feedback for our marketing people, try to reach out via Twitter/Mastodon/email or stop by the office.
> 
> If you have a support contract with Nextcloud GmbH, please report this problem via customer portal or email.

Nah, I'm just a sad user. Your marketing division is also doing a great job. The point is that let's say 90% of the development seems to be there just to deliver stuff to the marketing. Reliability and quality fall by the wayside, because you can hardly do marketing with that. ;) But no need to be pissed. It's simply criticism of a product, nothing more...

**Comment by kesselb** [comment]
> But no need to be pissed.

That is your interpretation. I'm just asking you to send your constructive feedback to the right people.

> I couldn't get it to work either.

I had a brief look with Nextcloud 27. Connection to an external SFTP storage with RSA private key works fine.

![image](https://user-images.githubusercontent.com/3902676/224404602-186c6380-dbaa-4430-9971-58b06ff9aff9.png)

1) Create a new keypair `ssh-keygen -f datadump2`. Leave the passphrase empty.
2) You see two files now: datadump2 (private key) and datadump2.pub (public key).
3) Copy the public key to the sftp server (the one you want to add in Nextcloud). I recommend ssh-copy-id to avoid issues with line endings `ssh-copy-id -i datadump2 datadump@none-of-your-business.com`. Enter the sftp password and continue.
4) Use `ssh -i datadump2 datadump@none-of-your-business.com` to check if the login works.
5) Add the external storage in Nextcloud. 
6) Have fun ;)

Note: We recently updated the ssh library. It's possible that older releases do not support modern keys ciphers.

**Comment by HannesJo0139** [comment]
Hmm weird, I am pretty sure it is exactly what I tried back then. However, thanks for your constructive reply. Having the exact procedure here may help if anyone else struggles with the gui issues.
